File: Data Structures & Algorithms/trapping-rain-water/submission-7.java
class Solution {
    public int trap(int[] height) {
        int l,r,leftMax,rightMax;
        l=0;
        r = height.length-1;
        leftMax = height[l];
        rightMax = height[r];
        int res = 0;
        while(l<r){
            if(leftMax<rightMax){
                l+=1;
                leftMax = Math.max(leftMax,height[l]);
                res += leftMax - height[l];

            } else{
                r-=1;
                rightMax = Math.max(rightMax,height[r]);
                res += rightMax-height[r];
            }
        }
        return res;
        // [0  2   0   3   1   0   1   3   2   1]


        //                                     0



    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-7.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        vis,l,res = {},0,0
        for i in range(len(s)):
            c =s[i]
            if c in vis:
                l = max(l,vis[c]+1)
            res = max(res,i-l+1)
            vis[c] = i
        return res
        
        